### PR TITLE
Apple AArch64 assembly backend for the Pasta fields

### PR DESCRIPTION
**Pull request by dannywillems**
Upstreams the Apple AArch64 assembly backend for the Pasta field arithmetic
from https://github.com/zakura-core/libraries.

> **Stacked on #102.** The base of this pull request is
> `dw/zizmor-and-pin-actions`, not `main`, so review #102 first and merge it
> first; this one then retargets to `main` automatically. Stacking was chosen
> over rebasing because both branches add jobs to `ci.yml` and
> `lints-stable.yml`, and this way the new jobs are written against the pinned,
> hardened versions of those files rather than being pinned after the fact.

**This pull request has been reduced in scope.** It previously carried eleven
commits; it now carries seven, stopping at "Document AArch64 Pasta field assembly",
plus a CHANGELOG entry and a CI job. The later assembly work is deferred to a follow-up so that
this one can be reviewed on a much smaller and lower-risk surface. Three of the
four concerns the earlier description raised no longer apply; see "What was
dropped" below.

The backend is behind an opt-in `aarch64-asm` feature and compiles assembly
only on `target_arch = "aarch64"` plus `target_vendor = "apple"`. A default
build is unchanged in what it computes.

Four of the seven commits are ports and keep their original author, date and
message, with an upstreaming block appended carrying the `format-patch` command
that reproduces each one and its `Source-Commit:` trailer. One is written here:
the manifest change, because the standard exclusions drop `Cargo.toml`. A
CHANGELOG entry is in its own commit.

## What a reviewer should look at

**1. A build script now runs for every consumer.** The manifest gains
`build = "build.rs"` and an optional `cc` build dependency. With the feature
off, `build.rs` compiles to a single `rerun-if-changed` line and `cc` is not in
the graph, but the build script itself runs unconditionally. That is a change
to the build surface of a widely depended on crate, and it is worth deciding on
deliberately rather than by patch review. It is the one concern from the
original description that survives the reduction.

**2. Assembly and `unsafe`.** `src/fields/aarch64_asm.rs` has three `unsafe`
blocks and one `extern "C"` declaration against the vendored
`src/asm/pasta_mul-armv8.S`. The FFI boundary is private to the crate. The
constant-time claim is stated in the module docs.

**3. `src/curves.rs` routes curve squaring through the field backend.** A
mechanical change, 18 lines each way, and the only file here that is not
AArch64-specific.

**4. The final commit documents the assembly and changes no instruction.**
"Document AArch64 Pasta field assembly" shows as 418 insertions and 373
deletions in the `.S`, which looks alarming for a comment change. It is not.
Normalising both revisions by stripping `//` comments and collapsing
whitespace leaves 329 instruction lines on each side, and the only four
differences are a space after a comma (`ldr x9, [x2]` against `ldr x9,[x2]`).
The assembler sees the same program. It is included because the assembly is
the part of this pull request that is hardest to review, and the comments are
what make it reviewable.

## What was dropped, and why it matters to the review

These were the largest risks in the original eleven commits, and none of them
is in this pull request any more:

* **`pow_vartime` is no longer touched at all.** The earlier series rewrote it
  for every target, not just Apple, so an error would have been a wrong field
  inverse on every platform. `git diff` against `main` now contains zero
  occurrences of `pow_vartime`.
* **The canonical-operand precondition is gone.** "Drop dead Montgomery carry
  limb" made the inline multiply depend on its right-hand operand already being
  canonical, checked only by `debug_assert!`. That commit is not here.
* **The move of multiplication and squaring into inline `asm!` is deferred**,
  which is what takes `unsafe` from five blocks to three and `extern "C"` from
  three declarations to one.
The CI job for the assembly backend is NOT dropped; see below.

## CI coverage for the new feature

Nothing outside a macOS arm64 runner compiles this assembly, so two jobs are
added against #102's already-pinned workflows:

* `aarch64-asm` in `ci.yml` builds and tests with the feature pinned on, and
  joins the required status checks.
* `clippy-aarch64-asm` in `lints-stable.yml`, because a lint inside gated code
  is invisible to the existing Linux clippy job.

Both guard against passing VACUOUSLY, which is the real risk with a
cfg-gated backend:

* the job asserts `uname -sm` is `Darwin arm64`, so if `macos-latest` ever
  moves back to x86_64 the job fails instead of going green without compiling
  any assembly;
* it re-runs the two differential tests by exact name and requires the output
  to say `2 passed`, because a test that compiles out still reports success.

Verified locally that the exact invocation reports `2 passed` with the feature
on, and does not report it with the feature off. `zizmor --offline` reports no
findings on the resulting workflows.

## Verification

On `aarch64-apple-darwin`, all green: `cargo build`; `cargo test`;
`cargo test --all-features`; `cargo test --no-default-features`;
`cargo test --features aarch64-asm`;
`cargo clippy --all-targets --all-features -- -D warnings`;
`cargo clippy --all-targets --features aarch64-asm -- -D warnings`;
`cargo fmt --check`.

The feature is genuinely exercised rather than merely compiled: enabling
`aarch64-asm` takes the unit tests from 29 to 31, and the build script emits
the assembled `pasta_mul-armv8.o`.

### CODE REVIEW

File: .github/workflows/lints-stable.yml [hunk 1]
```diff
@@ -16,3 +16,14 @@
...
+  clippy-aarch64-asm:
+    name: Clippy (Apple AArch64 assembly backend)
+    timeout-minutes: 30
+    # The gated backend is compiled out everywhere else, so the ubuntu
+    # clippy job above never sees it.
+    runs-on: macos-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Run clippy
+        run: cargo clippy --all-targets --features aarch64-asm -- -D warnings
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply .github/workflows/lints-stable.yml:42]: Fixed in 85017f6: the new `clippy-aarch64-asm` job now declares `permissions: contents: read`.

Set per job rather than for the whole workflow on purpose. The existing `clippy` job posts a check run through `actions-rs/clippy-check` with the `GITHUB_TOKEN`, so a workflow-wide `contents: read` default would take away the `checks: write` it needs. The new job only reads the checkout and runs cargo, so read is the whole scope it wants.

The other job this PR adds, `aarch64-asm` in `ci.yml`, needs nothing: that workflow already sets `permissions: contents: read` / `packages: read` at the top level and the job inherits it.

**Review by dannywillems — COMMENTED** [review]
Self utACK - run the port-diff.sh script.

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed 0e57e5bb4c17035fb0514ad5e360ca4032f943a7.

I did not review the assembly code directly, but I reviewed the tests that claim to check its consistency with the portable Rust code, and they look correct.

I built an assembly mutation testing harness (#107, stacked on this PR). It revealed that the following mutations do not trigger test failures, meaning that this PR is missing test coverage:

```
Result: 234 killed, 0 crashed, 0 invalid, 24 survived
asm-mutants: invalid mutants:

surviving mutants:
  adcs/5/carry-clear (line 73)
  adc/3/carry-clear (line 86)
  adcs/16/carry-clear (line 115)
  adcs/19/carry-clear (line 126)
  adcs/27/carry-clear (line 157)
  adcs/30/carry-clear (line 168)
  adc/11/carry-clear (line 170)
  adcs/38/carry-clear (line 198)
  adc/15/carry-clear (line 207)
  adc/19/carry-clear (line 302)
  adc/21/carry-clear (line 322)
  sbcs/9/carry-clear (line 365)
  sbcs/10/carry-clear (line 366)
  sbcs/11/carry-clear (line 367)
  csel/9/select-first (line 370)
  csel/10/select-first (line 371)
  csel/11/select-first (line 372)
  csel/12/select-first (line 373)
  adcs/60/carry-clear (line 404)
  adcs/65/carry-clear (line 424)
  adc/24/carry-clear (line 426)
  adc/25/carry-clear (line 433)
  adc/27/carry-clear (line 453)
  adcs/75/carry-clear (line 464)
```

I'm not going to block this PR on adding full test coverage (because it's quite possible that some of these might be e.g. 2^-32 edge cases), but we should go through these mutants and triage them before merging (to determine if there are any easy wins).

File: src/asm/pasta_mul-armv8.S [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    // q * p[2] is zero because p[2] = 0.
+    lsl x17,x3,#62          // x17 = low(q * p[3]).
+    // Carry from x10 + low(q*p[0]) is one exactly when x10 is nonzero.
+    subs xzr,x10,#1         // Set that cancellation carry.
+    umulh x14,x5,x3         // x14 = high(q * p[0]).
+    adcs x11,x11,x15        // Add low(q * p[1]) and cancellation carry.
+    umulh x15,x6,x3         // x15 = high(q * p[1]).
+    adcs x12,x12,xzr        // Propagate carry across zero p[2].
+    // high(q * p[2]) is zero.
+    adcs x13,x13,x17        // Add low(q * p[3]) and carry.
```
Comment (str4d) [inline src/asm/pasta_mul-armv8.S]: Mutant `adcs/60/carry-clear (line 404)` can be killed by testing with `t = [1, u64::MAX, u64::MAX, 0]`.

File: src/asm/pasta_mul-armv8.S [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+    subs x14,x10,x5         // Tentative result limb 0 = candidate - p[0].
+    sbcs x15,x11,x6         // Tentative result limb 1 minus p[1].
+    sbcs x16,x12,x7         // Tentative result limb 2 minus p[2] = 0.
+    sbcs x17,x13,x8         // Tentative result limb 3 minus p[3].
+
+    // If subtraction borrowed (`lo`), keep the original candidate.
+    csel x10,x10,x14,lo     // Select canonical integer limb 0.
+    csel x11,x11,x15,lo     // Select canonical integer limb 1.
+    csel x12,x12,x16,lo     // Select canonical integer limb 2.
+    csel x13,x13,x17,lo     // Select canonical integer limb 3.
```
Comment (str4d) [inline src/asm/pasta_mul-armv8.S]: The following mutants can be killed by exercising `from_mont(p)`:
- `csel/9/select-first (line 370)`
- `csel/10/select-first (line 371)`
- `csel/12/select-first (line 373)`

(Not `csel/11/select-first (line 372)` because `p[2] = 0`, so limb 2 subtraction is 0 regardless.)

File: src/fields/fp.rs [hunk 4]
```diff
@@ -369,6 +369,48 @@ impl Fp {
         Fp::montgomery_reduce(u[0], u[1], u[2], u[3], u[4], u[5], u[6], u[7])
     }
 
+    #[inline]
+    fn mul_runtime(&self, rhs: &Self) -> Self {
+        #[cfg(all(
+            feature = "aarch64-asm",
+            target_arch = "aarch64",
+            target_vendor = "apple"
```
Comment (dannywillems) [inline src/fields/fp.rs:383]: See comment https://github.com/zcash/pasta_curves/pull/107#discussion_r3929886860: we might want to drop the `vendor` configuration - this might be misleading.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,13 @@ and this project adheres to Rust's notion of
 [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Added
+- An assembly implementation of Pallas and Vesta base and scalar field
+  multiplication, squaring and Montgomery reduction for Apple AArch64, behind
```
Comment (dannywillems) [inline CHANGELOG.md:11]: Same comment as above, see https://github.com/zcash/pasta_curves/pull/107#discussion_r3929886860 for the "Apple" reference.

**Comment by daira** [comment]
## Documentation and assurance review

### The operand contract of the assembly `mul`, and how `from_u512` calls it

Reviewed 8ad85e9fab7929f6236960e472f432a4bd9ccd74 including the assembly, using an instruction-level model of the `.S` (see next comment) checked against the real binary on Apple M-series.

The wrapper's docstring says `mul` "multiplies two canonical Montgomery residues" ([aarch64_asm.rs L27](https://github.com/zcash/pasta_curves/blob/8ad85e9fab7929f6236960e472f432a4bd9ccd74/src/fields/aarch64_asm.rs#L27)), while `from_u512` ([fp.rs L301](https://github.com/zcash/pasta_curves/blob/8ad85e9fab7929f6236960e472f432a4bd9ccd74/src/fields/fp.rs#L301), [fq.rs L301](https://github.com/zcash/pasta_curves/blob/8ad85e9fab7929f6236960e472f432a4bd9ccd74/src/fields/fq.rs#L301)) multiplies an unreduced 256-bit half by `R2` or `R3` through `Mul`, so the unreduced value is the assembly's left operand. That gap is worth closing, because the routine's tolerance for an unreduced operand is asymmetric:

- A canonical left operand with any 256-bit right operand is always fine: in every round `t + lhs * rhs[i] < 2p + 2^64 * p < 2^320`, and the final candidate is below `2p`. Semolina's own wrappers call it this way round, with the constant on the left.
- An unreduced left operand is fine only while every right-operand limb in positions 1 to 3 stays a few units below `2^64`. The five-limb accumulator's fold of the high cross-products ends in `adc x23,x23,x17` ([L106](https://github.com/zcash/pasta_curves/blob/8ad85e9fab7929f6236960e472f432a4bd9ccd74/src/asm/pasta_mul-armv8.S#L106), [L148](https://github.com/zcash/pasta_curves/blob/8ad85e9fab7929f6236960e472f432a4bd9ccd74/src/asm/pasta_mul-armv8.S#L148), [L190](https://github.com/zcash/pasta_curves/blob/8ad85e9fab7929f6236960e472f432a4bd9ccd74/src/asm/pasta_mul-armv8.S#L190)), which discards its carry, and with `lhs[3] = 2^64 - 1` and `rhs[i] = 2^64 - 1` it wraps. Witness, both fields, `M = 2^64 - 1`, limbs little-endian: `lhs = [0, M, M, M]`, `rhs = [M, M, 0, 0]`; the pair satisfies `lhs * rhs < 2^256 * p`. The correct Pallas product `lhs * rhs * 2^-256 mod p` has limb 2 `0xda71d82f60a84f6d`; the backend returns `0xda71d82f60a84f6c`, low by `2^128` and still below `p`. For Vesta the correct limb 2 is `0xf5b20ea460e4c3a8` and the backend returns `0xf5b20ea460e4c3a7`. The portable `Fp::mul` agrees with the mathematical value on both.

As this PR stands the wrap is unreachable: `R2` and `R3` are the only right operands that ever meet an unreduced left operand, and none of their limbs 1 to 3 is above `2^64 - 4` in either field. But nothing in the crate states or checks either fact, and the documented contract is not the one `from_u512` relies on.

For context, the four ported commits correspond to the state of zakura-core/common's `main` between the merge of zakura-core/common#6 (13 August) and zakura-core/common#66 (20 August). zakura-core/common#66 ([ce54a938](https://github.com/zakura-core/common/commit/ce54a9384009555babb771df9eb612804d5f7443)) identified this, stated the contract in the module doc ([L24-L43](https://github.com/zakura-core/common/blob/ce54a9384009555babb771df9eb612804d5f7443/pasta_curves/src/fields/aarch64_asm.rs#L24-L43)), routed `from_u512` through the portable multiplication ([fp.rs L300-L306](https://github.com/zakura-core/common/blob/ce54a9384009555babb771df9eb612804d5f7443/pasta_curves/src/fields/fp.rs#L300-L306)), and added `aarch64_asm_mul_unreduced_lhs_matches_portable`, which asserts the `R2`/`R3` limb bound. Since that PR also moved the routines to inline `asm!` and was squash-merged, this PR's cut-off point left those pieces behind. They apply unchanged to the FFI routines here.

Suggested for this PR, as separate commits, independent of the inline-`asm!` move and of zakura-core/common#132's carry-limb drop (which the description already leaves out):

1. State the contract in the wrapper docstrings: canonical `lhs`, any 256-bit `rhs`; canonical input for `square` and `from_mont`.
2. Keep unreduced values off the assembly's left operand: carry over zakura-core/common#66's `from_u512` reroute, or write `R2 * d0 + R3 * d1`.
3. Carry over the limb-bound test, and add the witness as a test of the mathematical value on the portable path (tests below; the assembly variants fail on this head and document the boundary, so they belong with the fix rather than as they are).
4. A `debug_assert!` on the left operand's canonicity in the wrapper.

Separately, I am formalizing the three routines against an instruction-level model in CompElliptic, which will settle the exact wrap boundary and the `square`/`from_mont` range claims. #108's operand-range text should follow that or adopt zakura-core/common#66's wording; comment there.

The calling-convention side looks right: `mul` saves `x19`-`x24`, `sqr` saves `x19`-`x22`, the shared helper touches only caller-saved registers, frames are 16-byte aligned, and the `paciasp`/`autiasp` words are hint-space NOPs on non-arm64e targets. `build.rs` running for every consumer is fine by me, recorded in the changelog entry.

Reproduction files, attached below: `asmsim.py`, an instruction-level simulator of the `.S` (`python3 asmsim.py src/asm/pasta_mul-armv8.S` checks the model against the mathematical product on canonical and on unreduced inputs, then searches the `lhs = [x, M, M, M]`, `rhs = [M, M, 0, 0]` family and prints a witness per field); and `pasta-aarch64-asm-mul-witness-test.patch`, the four tests against 8ad85e9f (two per field, portable and assembly checked separately against the hard-coded value; `cargo test --release --features aarch64-asm accumulator_wrap_witness` gives 2 passed and 2 failed on Apple M-series).

### Witness tests

Fp; Fq is the same with the expected limbs given above.

```rust
/// An operand pair inside `lhs * rhs < 2^256 * p`, with `lhs` unreduced as `from_u512`
/// passes it, and the Montgomery product `lhs * rhs * 2^-256 mod p` computed independently.
#[cfg(test)]
const MUL_ACCUMULATOR_WRAP_WITNESS: ([u64; 4], [u64; 4], [u64; 4]) = (
    // lhs = 2^256 - 2^64
    [0, u64::MAX, u64::MAX, u64::MAX],
    // rhs = 2^128 - 1 < p
    [u64::MAX, u64::MAX, 0, 0],
    [0x11e964da0a175a5c, 0xabeacec71862cbc5, 0xda71d82f60a84f6d, 0x0a7fdd5bafc89a65],
);

#[test]
fn portable_mul_accumulator_wrap_witness() {
    let (lhs, rhs, expected) = MUL_ACCUMULATOR_WRAP_WITNESS;
    assert_eq!(Fp::mul(&Fp(lhs), &Fp(rhs)).0, expected);
}

#[cfg(all(feature = "aarch64-asm", target_arch = "aarch64", target_vendor = "apple"))]
#[test]
fn aarch64_asm_mul_accumulator_wrap_witness() {
    let (lhs, rhs, expected) = MUL_ACCUMULATOR_WRAP_WITNESS;
    assert_eq!(super::aarch64_asm::mul(&lhs, &rhs, &MODULUS.0, INV), expected);
}
```

Vesta expected limbs: `[0x5bd0d66d2601294f, 0x3f2f1d154ea8556b, 0xf5b20ea460e4c3a8, 0x347f03908050df13]`.

Posted with the assistance of Claude Fable 5.1

**Comment by daira** [comment]
Supporting file for the last comment: [`asmsim.py`](https://github.com/user-attachments/files/31961615/asmsim.py)

Run it with the local path to `pasta_curves/src/asm/pasta_mul-armv8.S` from the branch as an argument. The output is:
```
== Pallas Fp: inv = 0x992d30ecffffffff
  canonical random: 0 mismatches in 6000 calls
  lhs < 2^256, rhs < p, random: 0 mismatches in 2000 calls
  targeted lhs=[a0,M,M,M], rhs=[M,M,0,0]: 203 mismatches of 203
    witness lhs = 0xffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000
            rhs = 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff
            asm = 0x0a7fdd5bafc89a65da71d82f60a84f6cabeacec71862cbc511e964da0a175a5c
           math = 0x0a7fdd5bafc89a65da71d82f60a84f6dabeacec71862cbc511e964da0a175a5c
== Vesta Fq: inv = 0x8c46eb20ffffffff
  canonical random: 0 mismatches in 6000 calls
  lhs < 2^256, rhs < p, random: 0 mismatches in 2000 calls
  targeted lhs=[a0,M,M,M], rhs=[M,M,0,0]: 203 mismatches of 203
    witness lhs = 0xffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000
            rhs = 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff
            asm = 0x347f03908050df13f5b20ea460e4c3a73f2f1d154ea8556b5bd0d66d2601294f
           math = 0x347f03908050df13f5b20ea460e4c3a83f2f1d154ea8556b5bd0d66d2601294f
```

**Review by daira — CHANGES_REQUESTED** [review]
NACK until my review comment is addressed.

**Comment by daira** [comment]
The formalization mentioned above is in progress as a draft PR: daira/CompElliptic#35. So far it has an instruction-level model of the `.S`, generated from the vendored file, with 1052 outputs of the real routines checked by the kernel against the transcription (including the unreduced-operand cases); the correctness proofs under the operand contracts stated there are the next units.

Posted with the assistance of Claude Fable 5.1

**Comment by dannywillems** [comment]
> [zakura-core/common#66](https://github.com/zakura-core/common/pull/66) ([ce54a938](https://github.com/zakura-core/common/commit/ce54a9384009555babb771df9eb612804d5f7443)) identified this, stated the contract in the module doc ([L24-L43](https://github.com/zakura-core/common/blob/ce54a9384009555babb771df9eb612804d5f7443/pasta_curves/src/fields/aarch64_asm.rs#L24-L43)), routed `from_u512` through the portable multiplication ([fp.rs L300-L306](https://github.com/zakura-core/common/blob/ce54a9384009555babb771df9eb612804d5f7443/pasta_curves/src/fields/fp.rs#L300-L306)), and added `aarch64_asm_mul_unreduced_lhs_matches_portable`, which asserts the `R2`/`R3` limb bound. Since that PR also moved the routines to inline `asm!` and was squash-merged, this PR's cut-off point left those pieces behind. They apply unchanged to the FFI routines here.

Note: the commits from !66 are expected to be moved. This patch does not include everything intentionally, see for instance https://github.com/zcash/pasta_curves/pull/106.

**Comment by dannywillems** [comment]
See https://rust-lang.github.io/rfcs/2873-inline-asm.html for the inline assembly.

File: src/fields/fp.rs [hunk 6]
```diff
@@ -1346,3 +1356,140 @@ fn aarch64_asm_mul_unreduced_lhs_matches_portable() {
...
+}
+
+#[cfg(all(
+    test,
+    feature = "aarch64-asm",
+    target_arch = "aarch64",
+    target_vendor = "apple"
+))]
+#[test]
+fn aarch64_asm_mul_unreduced_lhs_near_modulus_rhs_matches_portable() {
```
Comment (dannywillems) [inline src/fields/fp.rs]: Note: LGTM.

File: src/fields/aarch64_asm.rs [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+    false
+}
+
+/// Multiplies two Montgomery residues for a Pasta modulus. `rhs` must be
+/// canonical (debug-asserted; a violation yields an incorrect residue, see
+/// the module docs). `lhs` may be unreduced only if every `rhs` limb is at
+/// most `2^64 - 4`; see the module docs for the carry-chain bound behind
+/// this.
+#[inline(always)]
+pub(super) fn mul(lhs: &Limbs, rhs: &Limbs, modulus: &Limbs, inv: u64) -> Limbs {
```
Comment (dannywillems) [inline src/fields/aarch64_asm.rs:102]: Review note: https://github.com/daira/CompElliptic/pull/35 will need to be updated accordingly.

File: CHANGELOG.md (hunk 5 above)
Comment (dannywillems) [reply CHANGELOG.md:11]: See https://github.com/zcash/pasta_curves/pull/110

File: src/fields/fp.rs (hunk 4 above)
Comment (dannywillems) [reply src/fields/fp.rs:383]: See https://github.com/zcash/pasta_curves/pull/110

File: src/fields/aarch64_asm.rs [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+    debug_assert!(
```
Comment (dannywillems) [inline src/fields/aarch64_asm.rs:103]:
```
`lhs` may be unreduced only if every `rhs` limb is at
/// most `2^64 - 4`;
```

This is unchecked in this patch, only `rhs` is checked.

**Comment by dannywillems** [comment]
@str4d as discussed in the call, [7567e4b](https://github.com/zcash/pasta_curves/pull/100/commits/7567e4b303c3a826e28ee9e98fb82f3c3519c0c7) is the commit fixing the bug.

**Comment by dannywillems** [comment]
Note: to add more coverage, I would add these test vectors https://github.com/zcash/pasta_curves/pull/108/changes/2b5f9338b4adabedf0640fed122a9c5aa8f5482d, in a follow-up.

**Comment by daira** [comment]
**The formalization's operand contracts for `mul`: the debug assertion checks the wrong operand, and three statements are more conservative than the code**

Reviewed efc0c69533f491743162f3263acfb6c23603ad91. The formalization mentioned above (no longer in daira/CompElliptic#35; it will be published separately) now covers the inline `asm!` block of `mul` at this head, against an instruction-level model, and shows that the block returns the canonical Montgomery product under either of two operand contracts:

- A. `lhs < p` and any four-limb `rhs`;
- B. any four-limb `lhs`, and `rhs < p` with `rhs[1]`, `rhs[2]`, `rhs[3]` each at most `2^64 - 3`.

Under either contract `lhs * rhs < 2^256 * p`, so the final candidate `(lhs * rhs + m * p) / 2^256` is below `2p < 2^256`, and the carry out of its fourth limb, which 7567e4b drops, is `0`. Measured against these, the PR's contract is off in both directions: the `debug_assert!` does not catch the one way the block can go wrong, and three statements require more than the code needs.

1. **The assertion checks the wrong operand: canonical `rhs` is neither necessary nor sufficient.** The module doc ([L24-L28](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L24-L28), [L53-L59](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L53-L59)), the wrapper docstring ([L96-L100](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L96-L100)), and the `debug_assert!` ([L103-L106](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L103-L106)) make a canonical `rhs` the requirement, and L53-L59 say that for `rhs >= R - p` the dropped limb yields an incorrect residue. It is not sufficient: by contract B, an unreduced `lhs` also needs `rhs[1]`, `rhs[2]`, `rhs[3]` at most `2^64 - 3`, which nothing checks, so the one kind of call that can produce a wrong result (unreduced `lhs`, canonical `rhs` with one of those limbs equal to `2^64 - 1`; the witness pair above is one) passes the assertion. It is not necessary either: by contract A, a canonical `lhs` gives a correct canonical result for every four-limb `rhs`, including `rhs >= R - p`, since the candidate can reach `R` only when both operands are unreduced, so the assertion rejects calls that are fine. The check that matches the code is `is_canonical(lhs) || (is_canonical(rhs) && rhs[1..].iter().all(|&l| l <= u64::MAX - 2))`, which also answers the review note at L103. Contract A also means that `from_u512` does not need the portable multiplication: `R2 * d0` with the unreduced half as `rhs` is inside it (Semolina's own wrappers put the constant on the left).

2. **The per-limb bound is `2^64 - 3`, on limbs 1 to 3 only.** The module doc's bound of `2^64 - 4` on every `rhs` limb ([L26-L27](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L26-L27), [L36-L37](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L36-L37)) is sufficient but not tight: contract B establishes `2^64 - 3` for limbs 1 to 3, and `rhs[0]` is unconstrained (round 0 has no carried accumulator to wrap). The doc's own arithmetic agrees: with `acc4 <= 2` and a carry of at most 1, a high product of at most `2^64 - 4` fits, and `rhs_limb = 2^64 - 3` caps `high(lhs[3] * rhs_limb)` at `2^64 - 4`. Whether `2^64 - 2` can wrap is not settled by the proofs. The `R2`/`R3` limb test is unaffected (`2^64 - 4` is stricter).

3. **`from_mont` needs no canonical input.** Its docstring ([L538](https://github.com/zcash/pasta_curves/blob/efc0c69533f491743162f3263acfb6c23603ad91/src/fields/aarch64_asm.rs#L538)) asks for a canonical residue; for every four-limb input the output is canonical and equal to `value * 2^-256 mod p` (the shared helper's result is at most `p`, and the conditional subtraction removes the one excess case). Harmless as a requirement, but not one the code has.

The statements about `square` are outside what is proved so far.

Posted with the assistance of Claude Fable 5.1
